adapter/statsd/statsd.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statsd
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"text/template"
+	"time"
+
+	"github.com/cactus/go-statsd-client/statsd"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/duration"
+
+	"istio.io/mixer/adapter/statsd/config"
+	"istio.io/mixer/pkg/adapter"
+)
+
+type (
+	// builder holds a connection to the statsd server which is shared by all aspects, as well as the parsed
+	// templates for constructing statsd metric names.
+	builder struct {
+		adapter.DefaultBuilder
+		client statsd.Statter
+	}
+
+	aspect struct {
+		rate      float32
+		client    statsd.Statter
+		templates map[string]*template.Template // metric name -> template
+	}
+)
+
+var (
+	name = "statsd"
+	desc = "Pushes statsd metrics"
+	// conf encodes the default configuration for this adapter
+	conf = &config.Params{
+		Address:                   "localhost:8125",
+		Prefix:                    "",
+		FlushDuration:             &duration.Duration{Seconds: 0, Nanos: int32(300 * time.Millisecond)},
+		FlushBytes:                512,
+		SamplingRate:              1.0,
+		MetricNameTemplateStrings: make(map[string]string),
+	}
+)
+
+// Register records the builders exposed by this adapter.
+func Register(r adapter.Registrar) {
+	r.RegisterMetricsBuilder(newBuilder())
+}
+
+func newBuilder() *builder {
+	flushDuration, err := ptypes.Duration(conf.FlushDuration)
+	if err != nil {
+		// we panic because we're reading in our static default config
+		panic(fmt.Errorf("failed to parse default duration, %v, as a time.Duration", conf.FlushDuration))
+	}
+
+	client, err := statsd.NewBufferedClient(conf.Address, conf.Prefix, flushDuration, int(conf.FlushBytes))
```

Comment:
This always creates a connection to "localhost" during Registration, seems incorrect.
If your intent is to internally share  `client statsd.Statter` across all aspect that use the same `conf.Address`, then you will need to implement a statd.Statter cache.
Connection should only be opened during call to `newaspect.`